src/core/cluster.py

```diff
@@ -134,16 +139,33 @@ def kafka(self) -> Kafka:
             substrate=self.substrate,
         )
 
+    @property
+    def clients(self) -> set[KarapaceClient]:
```

Comment:
**suggestion** Maybe we could add the passwords in this object such that it can be easily accessed. Right now it seems that passwords and clients are logically separated, clients here and passwords in a dictionary. I understand the two informations are in two different places (relation and peer), but I believe they "logically" belong to the same class.

Anyhow, It is not a strong point, up to you. I just believe it would help readability/structure.